### PR TITLE
UXPROD-3903: Upgrade MDES to version 3.0.0

### PR DESCRIPTION
**Pull request by nhanaa**
[UXPROD-3903](https://issues.folio.org/browse/UXPROD-3903): This PR will merge team @folio-org/bama's changes in MDES to support the new bursar export configuration feature. It includes changes to adjust to the new bursar export job schemas and util to convert legacy bursar export jobs to new jobs. This PR must be merged as the same time as `mod-data-export-worker`.

> [!IMPORTANT]
> https://github.com/folio-org/folio-export-common/pull/64 must be merged and `folio-export-common` pointed back to `master` before this PR is merged.
>
> The following must be merged before or at the same time as this PR, otherwise snapshot will break:
> - https://github.com/folio-org/ui-circulation-log/pull/204
> - https://github.com/folio-org/ui-export-manager/pull/165
> - https://github.com/folio-org/ui-organizations/pull/606
> - https://github.com/folio-org/ui-plugin-bursar-export/pull/108 OR https://github.com/folio-org/ui-plugin-bursar-export/pull/109
> - https://github.com/folio-org/mod-bulk-operations/pull/176
> - https://github.com/folio-org/mod-data-export-worker/pull/485

### CODE REVIEW

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ folio:
     schemaName: mod_data_export_spring_quartz
     changeLog: classpath:db/changelog/changelog-quartz.xml
 server:
-  port: 8081
+  port: 8082
```
Comment (khandramai) [inline src/main/resources/application.yml:18]: @nhanaa why did you change this port?

File: src/test/java/org/folio/des/InstallUpgradeIT.java [hunk 2]
```diff
@@ -87,7 +89,7 @@ class InstallUpgradeIT {
     new GenericContainer<>(
       new ImageFromDockerfile("mod-data-export-spring").withFileFromPath(".", Path.of(".")))
     .withNetwork(NETWORK)
-    .withExposedPorts(8081)
+    .withExposedPorts(8082)
```
Comment (khandramai) [inline src/test/java/org/folio/des/InstallUpgradeIT.java:92]: @nhanaa , changing 8081 -> 8082 required here as far as you changed this port in application.properties.

**Comment by khandramai** [comment]
@nhanaa , the problem in unit tests (container failed to start) is in changing application port from 8081 to 8082. I set the port values to be the same in application.properties and MOD_DES in the InstallUpgradeIT and all the tests are PASSED. So, please, consider possibility to return this value back to 8081 in both application.properties and InstallUpgradeIT class.

**Comment by khandramai** [comment]
@nhanaa , please, fix code smells and code duplication. After that code can be reviewed by @folio-org/firebird-backend .

File: src/test/java/org/folio/des/scheduling/quartz/QuartzExportJobSchedulerTest.java [hunk 3]
```diff
@@ -71,8 +72,9 @@ void testSchedule() throws SchedulerException {
     assertEquals(1, schedulerListener.getJobsAddedCount());
     assertEquals(1, schedulerListener.getJobsScheduledCount());
 
-    await().pollDelay(1, TimeUnit.SECONDS).timeout(10, TimeUnit.SECONDS).untilAsserted(
-      () -> assertEquals(1, jobListener.getJobsExecutedCount()));
+    // TODO - Should be reworked because this code leads to the test failure due to unstable result
```
Comment (khandramai) [inline src/test/java/org/folio/des/scheduling/quartz/QuartzExportJobSchedulerTest.java:75]: Consider changing this validation. It fails from time to time.

File: src/main/resources/application.yml (hunk 1 above)
Comment (nhanaa) [reply src/main/resources/application.yml]: I think I did it for my local environment to run before and accidentally pushed the change up. I have changed the ports back to 8081.

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,3 @@
 [submodule "folio-export-common"]
 	path = folio-export-common
-	url = https://github.com/folio-org/folio-export-common.git
```
Comment (ncovercash) [inline .gitmodules:3]: This should be reverted once bama's changes are merged

File: descriptors/ModuleDescriptor-template.json [hunk 5]
```diff
@@ -41,116 +45,64 @@
       "version": "1.0",
       "handlers": [
         {
-          "methods": [
```
Comment (ncovercash) [inline descriptors/ModuleDescriptor-template.json:44]: please undo autoformatted changes

File: pom.xml [hunk 6]
```diff
@@ -256,6 +262,19 @@
       <scope>test</scope>
     </dependency>
 
+    <!-- <dependency>
```
Comment (ncovercash) [inline pom.xml:265]: remove these, rather than commenting out

File: src/test/java/org/folio/des/service/config/impl/ExportTypeBasedConfigManagerTest.java [hunk 7]
```diff
@@ -26,24 +34,23 @@
 import org.folio.des.domain.dto.ModelConfiguration;
 import org.folio.des.domain.exception.RequestValidationException;
 import org.folio.des.scheduling.bursar.BursarExportScheduler;
+import org.folio.des.scheduling.bursar.BursarExportScheduler;
```
Comment (ncovercash) [inline src/test/java/org/folio/des/service/config/impl/ExportTypeBasedConfigManagerTest.java:37]: should this and the `org.quartz.Scheduler` be duplicated?

**Review by ncovercash — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/des/service/impl/JobServiceImpl.java [hunk 8]
```diff
@@ -197,41 +244,79 @@ public org.folio.des.domain.dto.Job upsertAndSendToKafka(org.folio.des.domain.dt
...
+      jobs.stream().filter(j -> !exportTypes.contains(j.getType())).toList()
+    );
   }
 
-  private List<Job> filterJobsMatchingExportTypes(List<Job> jobs, Set<ExportType> exportTypes) {
-    return jobs.stream()
+  private List<Job> filterJobsMatchingExportTypes(
+    List<Job> jobs,
+    Set<ExportType> exportTypes
+  ) {
```
Comment (azizbekxm) [inline src/main/java/org/folio/des/service/impl/JobServiceImpl.java:317]: preview version of parameter list was readable and concise

File: src/main/java/org/folio/des/service/impl/JobServiceImpl.java [hunk 9]
```diff
@@ -250,60 +335,47 @@ public void deleteJobs(List<Job> jobs) {
...
-      throw new NotFoundException(String.format("The URL of the exported file is missing for jobId: %s", job.getId()));
+      log.error(
+        "The URL of the exported file is missing for jobId={}.",
+        job.getId()
+      );
+      throw new NotFoundException(
+        String.format(
+          "The URL of the exported file is missing for jobId: %s",
+          job.getId()
+        )
```
[the remark this answers is not in the document]
Comment (azizbekxm) [reply src/main/java/org/folio/des/service/impl/JobServiceImpl.java:352]: If there is acceptable reason, please provide or revert back refactoring for format style to one line

**Comment by azizbekxm** [comment]
@nhanaa You have made a lot of changes relate to extend one line to multiple line method parametrs, tests and etc., Please provide some reason or revert all refactored these places. These changes doesn't look good to me and may make harder to read actual logic of code..

File: src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.folio.de.entity.BaseJob;
+import org.folio.des.domain.dto.LegacyExportTypeSpecificParameters;
+import org.hibernate.annotations.Type;
+
+@Entity
+@Table(name = "job")
+@Data
+public class LegacyJob extends BaseJob {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java:15]: For example we use this table for edifact export, and that jobs for us is not legacy jobs, and our parameters is not legacy.
I think we can simplify this. You can remain 1 class without this refactoring but introduce custom jsonb converter that can convert to generic jsonb to both dtos.

Here is example of usage:
`ublic class PersonCompanyJsonbConverter implements AttributeConverter<Object, Jsonb> {
 
    private static final Jsonb JSONB = JsonbBuilder.create();
 
    @Override
    public Jsonb convertToDatabaseColumn(Object object) {
        return JSONB.toJson(object);
    }
 
    @Override
    public Object convertToEntityAttribute(Jsonb jsonb) {
        JsonArray jsonArray = jsonb.fromJson(JsonArray.class);
        if (jsonArray.size() > 0) {
            JsonObject jsonObject = jsonArray.getJsonObject(0);
            if (jsonObject.containsKey("name")) {
                return JSONB.fromJson(jsonb.toString(), Person.class);
            } else if (jsonObject.containsKey("address")) {
                return JSONB.fromJson(jsonb.toString(), Company.class);
            }
        }
        throw new IllegalArgumentException("Cannot deserialize JSONB to Person or Company class.");
    }
}`


and how to use:
`@Entity
@Table(name = "my_data")
public class MyData {
 
    @Id
    private Long id;
 
    @Column
    @Convert(converter = PersonCompanyJsonbConverter.class)
    private Object data;
 
    // getters and setters
}`

File: src/main/java/org/folio/des/service/FolioTenantService.java [hunk 11]
```diff
@@ -55,6 +60,7 @@ protected void afterTenantUpdate(TenantAttributes tenantAttributes) {
       oldJobDeleteScheduler.scheduleOldJobDeletion(context.getTenantId());
       kafka.createKafkaTopics();
       kafka.restartEventListeners();
+      LegacyBursarMigrationUtil.recreateLegacyJobs(bursarExportLegacyJobService, jobService);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/des/service/FolioTenantService.java:63]: all other places in this method uses injected service, so code decoupled, testable, can be mocked etc.
Better to use service here as well instead of utility class.

File: src/main/java/org/folio/des/service/util/JobMapperUtil.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    result.setName(jobEntity.getName());
+    result.setDescription(jobEntity.getDescription());
+    result.setSource(jobEntity.getSource());
+    result.setIsSystemSource(jobEntity.getIsSystemSource());
+    result.setType(jobEntity.getType());
+    result.setExportTypeSpecificParameters(
+      jobEntity.getExportTypeSpecificParameters()
+    );
+    result.setStatus(jobEntity.getStatus());
+    if (
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/des/service/util/JobMapperUtil.java:28]: please use the same style code convention, statement should be on the same line after opening brackets

File: src/main/java/org/folio/des/service/util/JobMapperUtil.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+    result.setOutputFormat(jobEntity.getOutputFormat());
+    result.setErrorDetails(jobEntity.getErrorDetails());
+
+    return result;
+  }
+
+  public static org.folio.des.domain.dto.LegacyJob entityToDto(
+    LegacyJob legacyJobEntity
+  ) {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/des/service/util/JobMapperUtil.java:60]: the same with code formatting

File: src/test/java/org/folio/des/converter/DefaultExportConfigToModelConfigConverterTest.java [hunk 14]
```diff
@@ -4,15 +4,23 @@
...
 import java.util.UUID;
 
 import org.folio.des.config.JacksonConfiguration;
-import org.folio.des.domain.dto.BursarFeeFines;
+import org.folio.des.domain.dto.BursarExportFilter;
+import org.folio.des.domain.dto.BursarExportFilterAge;
+import org.folio.des.domain.dto.BursarExportFilterCondition;
+import org.folio.des.domain.dto.BursarExportFilterPatronGroup;
+import org.folio.des.domain.dto.BursarExportJob;
+// import org.folio.des.domain.dto.BursarFeeFines;
```
Comment (SerhiiNosko) [inline src/test/java/org/folio/des/converter/DefaultExportConfigToModelConfigConverterTest.java:18]: please delete commented code

File: src/test/java/org/folio/des/scheduling/quartz/QuartzExportJobSchedulerTest.java (hunk 3 above)
Comment (SerhiiNosko) [reply src/test/java/org/folio/des/scheduling/quartz/QuartzExportJobSchedulerTest.java:75]: in our recent prs where we introduced quartz - we did not see any issues with this test

File: src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java (hunk 10 above)
Comment (nhanaa) [reply src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java:15]: So if I understand it correctly, you want me to keep just one `Job` class instead of splitting it into `LegacyJob` and `Job` and use  a custom jsonb converter that can convert to generic jsonb to both dtos, right? I don't think I fully understand how you want me to change this code. Would we then have two different `ExportTypeSpecificParameters` attributes in the `Job` class?

File: src/main/java/org/folio/des/service/FolioTenantService.java (hunk 11 above)
Comment (nhanaa) [reply src/main/java/org/folio/des/service/FolioTenantService.java:63]: Should I then refactor `LegacyBursarMigrationUtil` to something like `BursarMigrationService` class?

Comment (SerhiiNosko) [reply src/main/java/org/folio/des/service/FolioTenantService.java:63]: yes, it would be better, in this case our components will be decoupled

File: src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java (hunk 10 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java:15]: we can use some abstart class for like BaseExportTypeSpecificParameters and in converter chose what instance to create:
ExportTypeSpecificParameters or LegacyExportTypeSpecificParameters based of content of jsonb field.
But if it hard to implement I am ok with this approach with mapped superclass.

Comment (nhanaa) [reply src/main/java/org/folio/de/entity/bursarlegacy/LegacyJob.java:15]: @SerhiiNosko I have tried implementing it as above like you said but I ran into problems with generating correct Java classes using OpenAPI generator for the abstract class and inherited classes like above. I have described it more in details in DM if you can take a look.

**Review by ncovercash — APPROVED** [review]

**Comment by khandramai** [comment]
@nhanaa , please, review Quality Gate issues.

**Comment by danetsao** [comment]
> @nhanaa , please, review Quality Gate issues.

Hi Slava, this was due to an issue with a test failing on Jenkins build, so I increased a few timeouts. Looks like all actions are passing, only one code smell relating to a List being parameterized. Let me know what you think about quality gate and the code smell.

**Comment by danetsao** [comment]
Here is the two screenshots of the Karate that these changes are related to tests passing. I will try to upload the e2e tests tommorow.
Screenshot 2024-01-31 at 2 51 46 PM
Screenshot 2024-01-31 at 2 51 56 PM

**Review by siarhei-charniak — APPROVED** [review]

**Comment by danetsao** [comment]
Here is a compressed folder with screenshots of all relevant karate tests passing. It was run with the following areas `common,testrail-integration,acquisitions,mod-audit,mod-bulk-operations,mod-bulk-edit`.
[UXPROD3903_Karate_Tests.zip](https://github.com/folio-org/mod-data-export-spring/files/14144395/UXPROD3903_Karate_Tests.zip)

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+            jobWithLegacyBursarParameters
+              .getExportTypeSpecificParameters()
+              .getVendorEdiOrdersExportConfig()
+          );
+        newJob
+          .getExportTypeSpecificParameters()
+          .setQuery(jobWithLegacyBursarParameters.getExportTypeSpecificParameters().getQuery());
+        newJob
+          .getExportTypeSpecificParameters()
+          .seteHoldingsExportConfig(
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java:100]: set?

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java [hunk 16]
```diff
@@ -0,0 +1,448 @@
...
+    }
+
+    for (JobWithLegacyBursarParameters jobWithLegacyBursarParameters : jobsToRecreate) {
+      log.info("job to recreate: {}", jobWithLegacyBursarParameters);
+      if (isLegacyJob(jobWithLegacyBursarParameters)) {
+        Job newJob = prepareNewJob();
+
+        newJob
+          .getExportTypeSpecificParameters()
+          .setVendorEdiOrdersExportConfig(
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java:90]: any reasons to touch edi, authrorities etc? as I understand you need only bursar jobs - and in previous implementation it was possible to schedule only one bursar job

Comment (NikitaSedyx) [reply src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java:90]: I mean to have only one config

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+
+  public void recreateLegacyJobs(
+    BursarExportLegacyJobService bursarExportLegacyJobService,
+    JobService jobService
+  ) {
+    log.info("searching for legacy jobs");
+    JobWithLegacyBursarParametersCollection response = bursarExportLegacyJobService.get(
+      0,
+      DEFAULT_LIMIT,
+      "status==SCHEDULED"
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java:67]: it looks a bit incorrect, job created at the moment when quartz reached scheduled time - you basically need to migrate configs (if required), but not scheduled jobs

one more questions, will previously executed jobs (completed or failed statuses) open correctly on UI?

Comment (ncovercash) [reply src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java]: All are now converted (jobs, old and new, as well as configs)

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java (hunk 16 above)
Comment (ncovercash) [reply src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java]: This has been resolved. You can verify the filtering logic in `BursarExportLegacyJobService`

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class BursarMigrationService {
+
+  private static final String BURSAR_EXPORT_MIGRATION_HEADER_CONSTANT = "LIB02";
+
+  // this ensures the currently scheduled configuration is updated and new jobs are created with the new schema
+  public void updateLegacyBursarConfigs(BurSarFeesFinesExportConfigService configService) {
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java:51]: BurSar?

File: src/main/java/org/folio/des/service/FolioTenantService.java [hunk 19]
```diff
@@ -55,6 +76,9 @@ protected void afterTenantUpdate(TenantAttributes tenantAttributes) {
       oldJobDeleteScheduler.scheduleOldJobDeletion(context.getTenantId());
       kafka.createKafkaTopics();
       kafka.restartEventListeners();
+
+      bursarMigrationService.updateLegacyBursarJobs(bursarExportLegacyJobService, jobService);
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/des/service/FolioTenantService.java:80]: this will be executed in all further releases even when already migrated? do you plan to remove this code or how will you manage migration process?

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java (hunk 18 above)
Comment (ncovercash) [reply src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java:51]: That's the old name of that existing class :(
I figured it'd be noisy to change it, so I opted to leave it alone

File: src/main/java/org/folio/des/service/FolioTenantService.java (hunk 19 above)
Comment (ncovercash) [reply src/main/java/org/folio/des/service/FolioTenantService.java:80]: IMO this should probably stay in place, at least for the foreseeable future.  All other types of migration actions (liquibase updates, RMB scripts, etc) are always checked on upgrade — there's nothing that prevents someone from updating this module from v1.0.0 to v999 or any other combination, and we should be able to handle any version range without breaking, meaning these migration scripts would need to stick around for some time.

**Review by khandramai — APPROVED** [review]

**Comment by khandramai** [comment]
LGTM. @danetsao , do you have any blockers to process with this?

**Comment by khandramai** [comment]
General remark - could you please apply FOLIO code style to all code because there are couple of places with incorrect code formatting.

**Comment by NikitaSedyx** [comment]
looks like it's breaking change, where is major increment here? in addition to this interfaces should be updated across module

File: src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java (hunk 18 above)
Comment (ncovercash) [reply src/main/java/org/folio/des/service/bursarlegacy/BursarMigrationService.java]: Renamed

File: src/main/java/org/folio/des/service/FolioTenantService.java [hunk 20]
```diff
@@ -67,4 +92,21 @@ protected void afterTenantDeletion(TenantAttributes tenantAttributes) {
     scheduledJobsRemover.deleteJobs(tenantId);
     oldJobDeleteScheduler.removeJobs(tenantId);
   }
+
+  protected boolean shouldUpdateBursar(TenantAttributes tenantAttributes) {
```
Comment (NikitaSedyx) [inline src/main/java/org/folio/des/service/FolioTenantService.java:96]: do I understand correctly that it will return true for 3.2.0 3.2.1 3.2.2 etc version?

why not to incapsulate everything in  bursarMigrationService and call only bursarMigrationService.migrate() in FolioTenantService? bursarMigrationService.migrate() will check everything and call required migrations

Comment (ncovercash) [reply src/main/java/org/folio/des/service/FolioTenantService.java:96]: Sure, I will rearrange that.
and yes, you're correct; I'm assuming this would be released in a 4.0.0 for Q

Comment (NikitaSedyx) [reply src/main/java/org/folio/des/service/FolioTenantService.java:96]: ah, ok, now I got it - missed that you use from version
